test/network_test.dart

```diff
@@ -0,0 +1,70 @@
+import 'dart:io';
+
+import 'package:http/http.dart';
+import 'package:nextcloud/src/http_client/http_client.dart';
+import 'package:nextcloud/src/network.dart';
+import 'package:test/test.dart';
+import 'package:mockito/mockito.dart';
+
+class HttpClientMock extends Mock implements HttpClient {}
+
+void main() {
+  group('Network', () {
+    final httpClientMock = HttpClientMock();
+    final httpRequest = Request('GET', Uri.https('test', ''));
+    const authString = 'authString';
+    const userAgent = 'dart-nextcloud';
+
+    setUpAll(() {
+      when(httpClientMock.send(httpRequest)).thenAnswer(
+        (_) async => StreamedResponse(
+          Stream.empty(),
+          HttpStatus.ok,
+        ),
+      );
+    });
+
+    test('should add defaultHeaders to request', () async {
+      final network = NextCloudHttpClient(
+        authString,
+        {
+          HttpHeaders.userAgentHeader: userAgent,
+        },
+        httpClientMock,
+      );
+
+      await network.send(httpRequest);
+
+      expect(httpRequest.headers[HttpHeaders.userAgentHeader], userAgent);
+    });
+
+    test('should not override library headers', () async {
```

Comment:
This case should maybe print something (not the test but the library) to inform the user

Reply:
Agreed! I will make it throw an exception when `defaultHeaders` attempt to override a library header.

Reply:
We have three types of headers:
- coreHeaders: set in `NextCloudHttpClient`
- defaultHeaders: set by users to be used if header does not exist
- requestHeaders: set either by the user or by the sub client, we do not know that on `NextCloudHttpClient` level.

I suggest the following order of importance: coreHeaders > requestHeaders > defaultHeaders

I would not throw in case that a requestHeader is overriding a defaultHeader since it could be a conscious user decision. In case that a defaultHeader is trying to override a coreHeader we can throw.

Or maybe a logging approach is in this case the better way to go?

Reply:
After giving this more thought I definitely think that logging is the better way, here. Whenever a header is overridden the library should log out a warning. This gives the whole header handling a more consistent touch.

The only question is how to integrate logging. Simplest way is by using just a callback that can be set by the user of the library. If no callback is set, logging is omitted. More complex alternative would be to pull in some kind of logging framework.

Reply:
I think using `assert`'s would also be an option, because the are not executed on release builds, but trigger in debug builds. They are already used in some places.

Reply:
Only for the case when trying to override coreHeaders, I guess?

Reply:
Yes